### ISSUE TITLE
use of `==` in tests

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#24878]
In #24771 there is some work to clean up tests. For linting and clarity purposes, tests like `assert a==b` could be passed to a function `test_eq(a, b): return a==b` so there would be only one place that uses `==` in a way that is intentionally testing the use of `==`. The docstring of that function could then tell when to use this function, why it is necessary, and add any linting directives as necessary -- in one place.

**Comment by asmeurer** [comment]
I don't agree. This just feels like unnecessary misdirection, which would make the tests harder to read. The arguments in #24771 don't really convince me. The tests that are using `== True` instead of `is True` are perfectly fine. One will only pass with `True` and the other will pass with `True` or `S.true`. This is no different from `assert x == 1` which will pass if `x` is `int(1)` or `Integer(1)`. That level of specificity (that a function returns a sympy type) really only needs to be tested once. The important part of the vast majority of tests is the value, not the type.

**Comment by smichr** [comment]
The reason I raise the point is that the desire to edit these away is happening again. There is insufficient documentation in the test itself to protect it. The "misdirection" would be a self-documenting preservation of the test in a specific form. I am under the impression that some tests using equality that aren't testing `== True` are intentional (perhaps like `a == a`). It is those that might benefit from a single testing function.

**Comment by asmeurer** [comment]
I see, so your suggestion is to only use this in places that look suspicious, not everywhere. 

Changing `==` to `is` where it's possible in the tests isn't bad, per se. All it effectively does is incorporate a type check alongside the value check. We should be clear though that this only works for those return objects that happen to be singletonized (both builtins and `S` objects). For everything else, we need an explicit `assert isinstance` or `assert type ==` in the tests to ensure the type is actually correct (like `x == 2` vs. `x == 2 and isinstance(x, Integer)`). This really only needs to be checked once per function. A type check doesn't actually need to be part of every test.

There's really only one test in the test suite that *needs* to be `== True` instead of `is True` or `is S.true`, and that's the test that `S.true == True` works. Tests that test `==` itself are fairly isolated and can be marked with a comment. Or they can even use `__eq__` directly if it feels necessary to detract future refactorers.

I kind of dislike that certain SymPy objects are singletonized at all, and that it's hard-coded as public API. It means that we can never make something like `x is Reals` stop working, even though SingleRegistry is an unnecessary overhead that provides no real benefit for it. But this is a separate issue from the tests, since the real problem is that these are documented this way.

And to be clear, I dislike changing `== True` to `is True` in symbolic code *outside* of the tests, for the reasons I described in that PR.

**Comment by asmeurer** [comment]
In general, I would like to just discourage these types of "refactoring" PRs that do nothing but churn the codebase for little overall benefit. They have a higher probability of introducing bugs than fixing them. 

Case in point: the above mentioned PR [changes a test](https://github.com/sympy/sympy/pull/24771#discussion_r1128783233) using `is` to `==` even though a comment specifically says not to. And it would have gone unnoticed due to the huge diff if this discussion hadn't made me think to check that specific test function. I haven't checked the diff to see if there are other such comments.

**Comment by asmeurer** [comment]
Also one final note here: these testing "gotchas" are the sort of thing I'd like to document in this document https://github.com/sympy/sympy/pull/23879/files#diff-7315bf5583c24ad4cf3af7e463383865180cf79ac52d21761a06f49544f8b2b2 (that PR has unfortunately not been merged yet even though it is more or less ready to go, but I hope to remedy that soon). For instance, I have a note there about using `unchanged(f, x)` instead of `assert f(x) == f(x)`.

**Comment by sylee957** [comment]
When we can introduce `unittest` https://docs.python.org/3/library/unittest.html

it can be more easier to subsume those in patterns like `assertTrue`, `assertFalse`, `assertNone`
than confusing the developers with sympy's three valued logic, which is not mainstream.

**Comment by bjodah** [comment]
I'm with @asmeurer on this one, PRs doing this kind of refactoring is an actively harmful form of bike-shedding since it drains the project of its most valuable asset: contributor's time (and our meta-discussion here is perhaps a case in point). Better to close such PRs and issues of that nature quickly, and with a friendly but firm explanation that the cost/benefit ratio isn't favorable, to save everyone from losing time.

**Comment by oscarbenjamin** [comment]
Some of those PRs are fine. The problem is when a new contributor does not really understand the code they are changing but that can be true of any kind of PR. As I said in gh-24508:

> It would interesting to see if these 1418 fixable things are changes that we actually want but again I'm not interested in a complete beginner submitting a PR with a diff running into thousands of lines. If someone wants to try the --fix option then they need to take time to review what exactly are the different things that ruff is doing and what the rules are. We probably don't want to apply all of those fixes and some judgment will be needed. Having someone apply the fix and dump the diff into a PR just pushes the work onto a reviewer. The PR would need an accompanying explanation and rationale which again makes this not really suitable for a beginner.

The idea of the change in gh-24771 is good but making the change is not a suitable task for a new contributor. Right now it is difficult to know exactly what is intended by `if x == True` whereas if it were changed to `if x is True`, `if x is S.true`, or `if x:` then the ambiguity would be removed. Especially in testing code the test should be as precise as possible so apart from actually testing `__eq__` itself we should generally use `is` where applicable and in fact even `__eq__` should be tested with `is True`:
```python
assert (a == b) is True  # test __eq__
```
Unfortunately the current ambiguity cannot be resolved unless someone who understands the distinction decides which case is supposed to apply to each of those lines of code. The codebase would be in a better shape by now had a linting rule like the one added there been in force in the past though.

Other cases like gh-24741 can be a lot easier because there is no ambiguity in how to change the code and the contributor in question clearly understood what they were doing and communicated the changes well.

**Comment by bjodah** [comment]
Maybe what I wrote was too much of a blanket statement, I was mostly referring to this kind of PRs when created by inexperienced developers. And gh-24741 does obviously not fit that description since the changes were accurate and made by a very skilled developer (we would be very lucky if the average new contributor would have a proficiency anywhere near that of @Skylion007).

And don't get me wrong I like that SymPy welcomes beginners, and many of us have spent considerable time teaching and tutoring. I simply feel that sometimes it would be in everybody's interest that some PRs got closed with some friendly feedback suggesting they start with something smaller.

**Comment by asmeurer** [comment]
The problem is that this kind of change is a trap for new contributors. They are nominally quite straightforward changes to make. The ruff --fix flag makes it even easier. The fact that we require a patch for GSoC also makes people want to find something simple to contribute. But changes like this are at best harmless to the codebase, but waste contributors time (like pointless stylistic changes), or worse, they contain mistakes that introduce new bugs. Changes that refactor tests are especially dangerous since there's the risk of doing something like deleting a test or modifying it to stop testing what it is supposed to. 

And the other thing that makes them a trap is that it's very often the case that there are subtleties that make the change much more complicated than it initially seems (like the subtleties between `== True`, `is True`, and `is true`). Here's an example of that: take https://github.com/sympy/sympy/issues/17944 (making functions use keyword-only arguments). Nominally this is very straight-forward: just add `*` to function signatures. But it's actually not. For one thing, you have to decide *which* arguments should be keyword-only, which requires deep thought about each individual API. And the change is strictly speaking backwards incompatible (which stalled my own attempts at fixing the issue https://github.com/sympy/sympy/pull/21803).

I would much rather new contributors focus on fixing actual bugs, rather than shifting code around that wasn't really buggy in the first place. Bug fixes are harder contributions to make, and they have the risk of being much harder than you expected. But there are also a lot of bugs in SymPy, and it shouldn't be too hard to find a reasonable contribution to make after a little looking. I would also mention that if someone is applying to GSoC, a bugfix PR is a much better signal to us that you are capable of doing their project than a cleanup/linting PR. New features can also be a good place to start if you feel up to it, especially those that focus on filling in gaps in existing functionality (e.g., adding code printers for missing functions).

**Comment by Sourabh5768** [comment]
Dear team,

I am excited to contribute to this project and offer my skills. Please let me support the team's efforts and collaborate effectively. Looking forward to working with you all.
